File: Session_10_StringMatching/multi_language_implementations/rabin_karp.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const d = 256
const q = 101

func runRabinKarp(pat, txt string) {
	M := len(pat)
	N := len(txt)
	p := 0
	t := 0
	h := 1

	for i := 0; i < M-1; i++ {
		h = (h * d) % q
	}

	for i := 0; i < M; i++ {
		p = (d*p + int(pat[i])) % q
		t = (d*t + int(txt[i])) % q
	}

	fmt.Println("Rabin-Karp Search details:")
	fmt.Printf("  Pattern hash: %d\n", p)

	for i := 0; i <= N-M; i++ {
		if p == t {
			fmt.Printf("  Hash match found at index %d (hash: %d). Checking characters...\n", i, t)
			j := 0
			for j = 0; j < M; j++ {
				if txt[i+j] != pat[j] {
					break
				}
			}
			if j == M {
				fmt.Printf("    Pattern found at index: %d\n", i)
			}
		}

		if i < N-M {
			t = (d*(t-int(txt[i])*h) + int(txt[i+M])) % q
			if t < 0 {
				t = t + q
			}
		}
	}
}

func main() {
	txt := "GEEKS FOR GEEKS"
	pat := "GEEK"

	fmt.Println("=== Rabin-Karp String Search ===")
	fmt.Printf("Text: \"%s\"\n", txt)
	fmt.Printf("Pattern: \"%s\"\n", pat)
	fmt.Println()

	runRabinKarp(pat, txt)

	fmt.Println()
	fmt.Print("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_10_StringMatching/multi_language_implementations/rabin_karp.rs
use std::io::{self, BufRead};

const D: i32 = 256;
const Q: i32 = 101;

fn run_rabin_karp(pat: &str, txt: &str) {
    let pat_bytes = pat.as_bytes();
    let txt_bytes = txt.as_bytes();
    let m = pat_bytes.len();
    let n = txt_bytes.len();
    
    let mut p = 0;
    let mut t = 0;
    let mut h = 1;
    
    for _ in 0..m - 1 {
        h = (h * D) % Q;
    }
    
    for i in 0..m {
        p = (D * p + pat_bytes[i] as i32) % Q;
        t = (D * t + txt_bytes[i] as i32) % Q;
    }
    
    println!("Rabin-Karp Search details:");
    println!("  Pattern hash: {}", p);
    
    for i in 0..=n - m {
        if p == t {
            println!("  Hash match found at index {} (hash: {}). Checking characters...", i, t);
            
            let mut j = 0;
            while j < m {
                if txt_bytes[i + j] != pat_bytes[j] {
                    break;
                }
                j += 1;
            }
            if j == m {
                println!("    Pattern found at index: {}", i);
            }
        }
        
        if i < n - m {
            t = (D * (t - txt_bytes[i] as i32 * h) + txt_bytes[i + m] as i32) % Q;
            if t < 0 {
                t += Q;
            }
        }
    }
}

fn main() {
    let txt = "GEEKS FOR GEEKS";
    let pat = "GEEK";
    
    println!("=== Rabin-Karp String Search ===");
    println!("Text: \"{}\"", txt);
    println!("Pattern: \"{}\"\n", pat);
    
    run_rabin_karp(pat, txt);
    
    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
